From cb445d95c26f5f3338fa8e496337a8de94ce155d Mon Sep 17 00:00:00 2001
From: Agarwal Ashish <ashishka@codeaurora.org>
Date: Fri, 17 Jan 2014 09:13:10 +0530
Subject: [PATCH 2/2] wlan: Prevent memory allocation when holding spin lock.

During APss(CPU3) power-collapse, WLAN Host Driver enter/exits
from IMPS. After entering/exiting from IMPS
hdd_conf_mcastbcast_filter() allocates memory
for wlanRxpFilterParam without unlocking spin lock.

CRs-Fixed: 601498
Change-Id: Ife3b3fe11b866e585ed6bbe562e49d38308e2176
---
 CORE/HDD/src/wlan_hdd_early_suspend.c | 16 ++++++++++++----
 1 file changed, 12 insertions(+), 4 deletions(-)

diff --git a/CORE/HDD/src/wlan_hdd_early_suspend.c b/CORE/HDD/src/wlan_hdd_early_suspend.c
index e8ea9db..fc274fe 100644
--- a/CORE/HDD/src/wlan_hdd_early_suspend.c
+++ b/CORE/HDD/src/wlan_hdd_early_suspend.c
@@ -1124,9 +1124,11 @@ static void hdd_PowerStateChangedCB
        pHddCtx->hdd_ignore_dtim_enabled = TRUE;
    }
    spin_lock(&pHddCtx->filter_lock);
-   if((newState == BMPS) &&  pHddCtx->hdd_wlan_suspended) {
+   if ((newState == BMPS) &&  pHddCtx->hdd_wlan_suspended)
+   {
       spin_unlock(&pHddCtx->filter_lock);
-      if (VOS_FALSE == pHddCtx->sus_res_mcastbcast_filter_valid) {
+      if (VOS_FALSE == pHddCtx->sus_res_mcastbcast_filter_valid)
+      {
           pHddCtx->sus_res_mcastbcast_filter =
               pHddCtx->configuredMcastBcastFilter;
           pHddCtx->sus_res_mcastbcast_filter_valid = VOS_TRUE;
@@ -1152,16 +1154,22 @@ static void hdd_PowerStateChangedCB
        * resume request will be lost. So reconfigure the filters on detecting
        * a change in the power state of the WCN chip.
        */
+      spin_unlock(&pHddCtx->filter_lock);
       if (IMPS != newState)
       {
+           spin_lock(&pHddCtx->filter_lock);
            if (FALSE == pHddCtx->hdd_wlan_suspended)
            {
+                spin_unlock(&pHddCtx->filter_lock);
                 hddLog(VOS_TRACE_LEVEL_INFO,
-                          "Not in IMPS/BMPS and suspended state");
+                       "Not in IMPS/BMPS and suspended state");
                 hdd_conf_mcastbcast_filter(pHddCtx, FALSE);
            }
+           else
+           {
+               spin_unlock(&pHddCtx->filter_lock);
+           }
       }
-      spin_unlock(&pHddCtx->filter_lock);
    }
 }
 
-- 
1.8.2.1

